LAYOUT_ENGINE.md

```diff
@@ -0,0 +1,177 @@
+
+<p align="center">
+  <img width="450" alt="kedro-viz-drawing" src="https://user-images.githubusercontent.com/45365769/117349544-67d75680-aea3-11eb-8c72-d1b88f4b930c.png">
+</p>
+
+We break down graph drawings like ours shown above into our essential and unique principles:
+
+- The nodes tend to be close together
+- The nodes do not overlap
+- The nodes align on each axis
+- The edges tend to the vertical
+- The edges do not cross
+- The edges point in the same direction
+- The trees have symmetry
+
+We form our [constraints](https://en.wikipedia.org/wiki/Constraint_(mathematics)) between nodes and edges using these rules, which allows us to frame the graph drawing problem as a [constrained optimisation problem](https://en.wikipedia.org/wiki/Constrained_optimization):
+
+> Given the graph _G_ contains set of nodes _N = { A, B, C… }_ and set of edges _E = { A → B, A → C… }_ and a set of constraints _C_
+> 
+> - for every node pair _{ A, B }_ in all nodes _N_
+>   - minimise position distance _|A<sub>p</sub> - B<sub>p</sub>|_
+>   - subject to all constraints in _C_
+> - for every edge _A → B_
+>   - draw path from _A<sub>p</sub>_ to _B<sub>p</sub>_ avoiding all nodes _N_
+
+In practice we've made two key simplifcations:
+
+- We consider node position variables in X and Y axes as [independent](https://en.wikipedia.org/wiki/Dependent_and_independent_variables), such that a node may freely change position in X without affecting its position in Y and vice versa
+- We consider the routing problem independent from the layout problem
+
+## The constraints
+
+From each rule we form [geometric constraints](https://en.wikipedia.org/wiki/Geometric_constraint_solving) as relations between our node positions:
+
+#### __Row Constraint__ 
+
+"Edges point in the same direction" and "Nodes do not overlap"
+
+>- for every edge _A → B_ in all edges _E_
+>    - ensure minimum distance and direction A to B in Y-axis  
+>    __B<sub>y</sub> ≥ A<sub>y</sub> + d__
+>    - where _A_ is the _source_ node and _B_ is the _target_ node
+>    - where _d_ is the minimum vertical row distance required
+
+_Notes_
+- we implemented this as our __row constraint__ ([code](https://github.com/quantumblacklabs/kedro-viz/blob/210974eb743a8116d6656ce41abe5ecee770dbcd/src/utils/graph/constraints.js#L9-L22))
+- and the __layer constraint__ by adding layer nodes ([code](https://github.com/quantumblacklabs/kedro-viz/blob/210974eb743a8116d6656ce41abe5ecee770dbcd/src/utils/graph/layout.js#L106-L136))
+- the expected performance is _O(E)_ [linear complexity](https://en.wikipedia.org/wiki/Computational_complexity) in number of edges
+- this equation forms a [linear constraint](https://en.wikipedia.org/wiki/Linear_programming) with one minimal solution
```

Comment:
Nit: I think the first letter of each bullet line should be capitalised, e.g.
```
- We implemented this as our __row constraint__ ...
```
This is because they read as complete sentences. The same goes for later sections.